code/controllers/subsystem/tts.dm

```diff
@@ -0,0 +1,227 @@
+	var/list/headers = list()
+	headers["Authorization"] = CONFIG_GET(string/tts_http_token)
+	request.prepare(RUSTG_HTTP_METHOD_GET, "[CONFIG_GET(string/tts_http_url)]/tts-voices", "", headers)
+	request.begin_async()
+	UNTIL(request.is_complete())
+	var/datum/http_response/response = request.into_response()
+	if(response.errored || response.status_code != 200)
+		return SS_INIT_FAILURE
+	available_speakers = json_decode(response.body)
+	tts_enabled = TRUE
+	rustg_file_write(json_encode(available_speakers), "data/cached_tts_voices.json")
+	rustg_file_write("rustg HTTP requests can't write to folders that don't exist, so we need to make it exist.", "tmp/tts/init.txt")
+	return SS_INIT_SUCCESS
+
+/datum/controller/subsystem/tts/proc/play_tts(target, sound, datum/language/language, local, range = 7)
+	if(local)
+		SEND_SOUND(target, sound)
+		return
+
+	var/turf/turf_source = get_turf(target)
+	if(!turf_source)
+		return
+
+	var/channel = SSsounds.random_available_channel()
+	var/listeners = get_hearers_in_view(range, turf_source)
+
+	for(var/mob/listening_mob in listeners | SSmobs.dead_players_by_zlevel[turf_source.z])//observers always hear through walls
+		var/datum/language_holder/holder = listening_mob.get_language_holder()
+		if(!listening_mob.client?.prefs.read_preference(/datum/preference/toggle/sound_tts))
+			continue
+
+		if(get_dist(listening_mob, turf_source) <= range && holder.has_language(language, spoken = FALSE))
+			listening_mob.playsound_local(
+				turf_source,
+				sound,
+				vol = listening_mob == target? 60 : 85,
+				falloff_exponent = SOUND_FALLOFF_EXPONENT,
+				channel = channel,
+				pressure_affected = TRUE,
+				sound_to_use = sound,
+				max_distance = SOUND_RANGE,
+				falloff_distance = SOUND_DEFAULT_FALLOFF_DISTANCE,
+				distance_multiplier = 1,
+				use_reverb = TRUE
+			)
+
+/datum/controller/subsystem/tts/proc/handle_request(list/entry)
+	var/timeout_time = entry[START_TIME_INDEX] + message_timeout
+	if(timeout_time < world.time)
+		cached_voices -= entry[IDENTIFIER_INDEX]
+		return
+	var/datum/http_request/request = entry[REQUEST_INDEX]
+	request.begin_async()
+	in_process_tts_messages += list(entry)
+
+/datum/controller/subsystem/tts/fire(resumed)
+	if(!tts_enabled)
+		flags |= SS_NO_FIRE
+		return
+
+	if(!resumed)
+		while(length(in_process_tts_messages) < max_concurrent_requests && queued_tts_messages.len > 0)
+			var/list/entry = popleft(queued_tts_messages)
+			handle_request(entry)
+		current_processing_tts_messages = in_process_tts_messages.Copy()
+
+	// For speed
+	var/list/processing_messages = current_processing_tts_messages
+	while(processing_messages.len)
+		var/current_message = processing_messages[processing_messages.len]
+		processing_messages.len--
+		var/datum/http_request/request = current_message[REQUEST_INDEX]
+		if(!request.is_complete())
+			continue
+
+		var/datum/http_response/response = request.into_response()
+		in_process_tts_messages -= list(current_message)
```

Comment:
how large do you expect this list to be? (the left one)

Reply:
it'll only ever contain `max_concurrent_requests` entries